File: lib/test/t_TabuList_std.cxx
#include <iostream>

#include <openturns/ClaytonCopulaFactory.hxx>
#include <openturns/ComposedCopula.hxx>
#include <openturns/FrankCopulaFactory.hxx>
#include <openturns/NormalCopulaFactory.hxx>
#include <openturns/RandomGenerator.hxx>

#include "otagrum/TabuList.hxx"

OT::Description DescrFromStringVect(const std::vector<std::string> &v)
{
  OT::Description res;
  for (auto s : v)
  {
    res.add(s);
  }
  return res;
}

OT::Sample getTrucBizarre(OT::UnsignedInteger size)
{
  OT::Collection<OT::Distribution> copulas;

  OT::UnsignedInteger dim = 3;

  OT::CorrelationMatrix R(dim);
  for (OT::UnsignedInteger i = 0; i < dim; i++)
    for (OT::UnsignedInteger j = 0; j < i; j++)
      R(i, j) = 0.64;

  auto copula1 = OT::FrankCopula(4.0);
  copula1.setDescription(DescrFromStringVect({"A1", "A2"}));
  auto copula2 = OT::ClaytonCopula(3.0);
  copula2.setDescription(DescrFromStringVect({"B1", "B2"}));

  copulas.add(copula1);
  copulas.add(copula2);
  copulas.add(OT::NormalCopula(R));

  auto copula = OT::ComposedCopula(copulas);
  return copula.getSample(size);
}

void tests(void)
{
  try
  {
    OT::RandomGenerator::SetSeed(0);
    OT::Sample sample = getTrucBizarre(4000);
    std::cout << "Sample size: " << sample.getSize() << std::endl
              << "Sample dimension:" << sample.getDimension() << std::endl
              << "Sample " << std::endl << sample << std::endl;
    {
      std::cout << "\n\n";
      OTAGRUM::TabuList learner(sample);
      std::cout << "go" << std::endl;
      auto ndag = learner.learnDAG();
      std::cout << "\n****\n"
                << "Named DAG"
                << "\n****\n"
                << ndag.__str__() << std::endl;
      std::cout << ndag.toDot() << std::endl;
    }
  }
  catch (const OT::Exception & e)
  {
    std::cout << e.__repr__() << std::endl;
  }
  catch (const gum::Exception & e)
  {
    GUM_SHOWERROR(e);
  }
};

std::string dirname()
{
  std::string res = __FILE__;
  const size_t last_slash_idx = res.find_last_of("\\/");
  if (std::string::npos != last_slash_idx)
  {
    res.erase(last_slash_idx + 1);
  }
  return res;
}

int main(void)
{
//   OT::Log::Show(OT::Log::ALL);
  tests();

  return EXIT_SUCCESS;
}

File: python/test/t_TabuList_std.py
#!/usr/bin/env python

import os
import openturns as ot
import otagrum


def generateDataForSpecificInstance(size):
    R = ot.CorrelationMatrix(3)
    R[0, 1] = 0.5
    R[0, 2] = 0.45
    collection = [ot.FrankCopula(3.0), ot.NormalCopula(R), ot.ClaytonCopula(2.0)]
    copula = ot.BlockIndependentCopula(collection)
    copula.setDescription("ABCDEFG")
    return copula.getSample(size)


def testSpecificInstance():
    size = 1000
    data = generateDataForSpecificInstance(size)
    learner = otagrum.TabuList(data)

    ndag = learner.learnDAG()
    idA = ndag.getDescription().find("A")
    idB = ndag.getDescription().find("B")
    idC = ndag.getDescription().find("C")
    idD = ndag.getDescription().find("D")
    idE = ndag.getDescription().find("E")
    idG = ndag.getDescription().find("G")

    dag = ndag.getDAG()
    assert dag.existsArc(idA, idB)
    assert dag.existsArc(idC, idD)
    # depends on map order ?
    assert dag.existsArc(idD, idE) or dag.existsArc(idE, idD)
    assert len(dag.children(idB)) == 0
    assert len(dag.children(idG)) == 0


def testAsiaDirichlet():
    data = ot.Sample.ImportFromTextFile(
        os.path.join(os.path.dirname(__file__), "asia_dirichlet_5000.csv"), ","
    )
    learner = otagrum.TabuList(data)
    learner.setVerbosity(True)
    dag = learner.learnDAG()
    print(dag.toDot())


testSpecificInstance()
testAsiaDirichlet()
